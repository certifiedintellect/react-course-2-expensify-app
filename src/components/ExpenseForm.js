import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';



export default class ExpenseForm extends React.Component{
 
    constructor(props){
        super(props)
        this.state = {
            description: props.expense? props.expense.description : '',
            note: props.expense ? props.expense.note: '',
            amount: props.expense ? (props.expense.amount).toString() : 0,
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calenderFocused: false,
            error: ''
        }
    }

   onDescriptionChange = (e) => {
       const description = e.target.value
       this.setState((state) => ({description}))
   }
   
   onNotesChange = (e) => {
       const note = e.target.value
       this.setState((state) => ({note}))
   }

   onAmountChange = (e) => {
       const amount = e.target.value

       if(amount.match(/^\d*(\.\d{0,2})?$/)){
           this.setState((state) => ({amount}))
       }

   }

   onDateChange = (createdAt) => {
       if(createdAt){
        this.setState((state) => ({createdAt}))
       }
   }

   onFocusChange = ({focused}) => {
       this.setState((state) => ({calenderFocused: focused}))
   }

   onFormSubmit = (e) => {
       const description = this.state.description
       const amount = this.state.amount

       e.preventDefault()

       if(!description || !amount){
            this.setState((state) => ({error: 'Please provide description and amount'}))
       } else{
           this.props.onSubmit({
               description: this.state.description,
               amount: parseFloat(this.state.amount, 10) ,
               createdAt: this.state.createdAt.valueOf(),
               note: this.state.note
           })
       }

   }

    render(){
        return (
            <div>
               {this.state.error && <p>{this.state.error}</p>}
               <form onSubmit={this.onFormSubmit}>
                  <input type="text"
                   placeholder="description"
                   value = {this.state.description}
                   onChange = {this.onDescriptionChange}
                   autoFocus />
                   <input type="text"
                   placeholder="amount"
                   value = {this.state.amount}
                   onChange = {this.onAmountChange}/>
                   <SingleDatePicker
                     date={this.state.createdAt} 
                     onDateChange = {this.onDateChange}
                     focused={this.state.calenderFocused}
                     onFocusChange={this.onFocusChange} 
                     numberOfMonths = {1}
                     isOutsideRange = {() => false}
                    />
                   <textarea placeholder="Enter the notes (optional)" 
                   value = {this.state.note}
                   onChange = {this.onNotesChange}/>
                   <button>Add Expense</button>
               </form>
            </div>
        )
    }
}

