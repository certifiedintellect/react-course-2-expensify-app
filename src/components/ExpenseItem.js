import React from 'react';
import {Link} from 'react-router-dom'

const ExpenseItem = ({id, description, amount, createdAt}) => (
    <div>
      <h3>{description} - <Link to={`/edit/${id}`}>Edit Expense</Link></h3>
      <p>{amount} - {createdAt}</p>
      
    </div>
)



export default ExpenseItem;