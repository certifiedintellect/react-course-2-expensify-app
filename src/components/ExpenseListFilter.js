import React from 'react';
import {connect} from 'react-redux';
import {setFilterText, sortByAmount, sortByDate, setStartDate, setEndDate} from '../actions/filters'

import {DateRangePicker} from 'react-dates';

class ExpensesListFilter extends React.Component{
    state = {
        calenderFocused: null
    }

    onDatesChange = ({startDate, endDate}) => {
        this.props.dispatch(setStartDate(startDate))
        this.props.dispatch(setEndDate(endDate))
    } 

    onFocusChange = (calenderFocused) => {
        this.setState((state) => ({calenderFocused}))
    }


    render(){
        return (
            <div>
            <input type="text" value={this.props.filters.text} onChange={(e) => {
                this.props.dispatch(setFilterText(e.target.value))
            }}/>
            <select value={this.props.filters.sortBy}
            onChange = {(e) => {
                if(e.target.value === 'date'){
                    this.props.dispatch(sortByDate())
                } else{
                    this.props.dispatch(sortByAmount())
                }
            }}>
               <option value="date">Date</option>
               <option value="amount">Amount</option>
            </select>
            <DateRangePicker
            startDate = {this.props.filters.startDate}
            endDate = {this.props.filters.endDate}
            onDatesChange = {this.onDatesChange}
            focusedInput = {this.state.calenderFocused}
            onFocusChange = {this.onFocusChange}
            numberOfMonths = {1}
            showClearDates = {true}
            isOutsideRange = {() => false}
            />
          </div>
        )
    }
}


const mapStateToProps = (state) => ({
    filters: state.filters
})

const ConnectedExpensesListFilter = connect(mapStateToProps)(ExpensesListFilter)

export default ConnectedExpensesListFilter