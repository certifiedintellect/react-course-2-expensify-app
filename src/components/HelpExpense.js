import React from 'react';

const HelpExpense = () => (
    <div>
       This is HelpExpenseComponent
    </div>
)


export default HelpExpense;