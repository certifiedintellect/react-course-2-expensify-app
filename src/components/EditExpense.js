import React from 'react';
import {connect} from 'react-redux'
import {editExpense, removeExpense} from '../actions/expenses'
import ExpenseForm from './ExpenseForm'

const EditExpense = (props) => {
  
    return (
      <div>
        This is EditExpenseComponent for {props.match.params.id}
        <ExpenseForm expense={props.expense}
        onSubmit = {(expense) => {
           props.dispatch(editExpense(props.match.params.id, expense))
           props.history.push('/')
        }}/>
        <button onClick={() => {
           props.dispatch(removeExpense({id:props.match.params.id}))
           props.history.push('/')
        }}>Remove</button>
     </div>
     )
  }
    

const mapStateToProps = (state, props) => (
  {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id 
    })
  }
)

export default connect(mapStateToProps)(EditExpense)