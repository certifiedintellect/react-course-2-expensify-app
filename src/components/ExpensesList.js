import React from 'react';
import {connect} from 'react-redux';
import ExpenseItem from './ExpenseItem';
import selectExpenses from '../selectors/expenses';

export const ExpensesList = (props) => (
    <div>
      <h3>Expenses List</h3>
      {
          props.expenses.map((expense) => {
              return <ExpenseItem key={expense.id} {...expense}/>
          })
      }
    </div>
)

const mapStateToProps = (state) => ({
    expenses: selectExpenses(state.expenses, state.filters)
})

const ConnectedExpensesList = connect(mapStateToProps)(ExpensesList)

export default ConnectedExpensesList