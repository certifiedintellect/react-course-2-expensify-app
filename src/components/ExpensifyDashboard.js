import React from 'react';
import ExpensesList from './ExpensesList'
import ExpenseListFilter from './ExpenseListFilter';

const ExpensifyDashboard = () => (
    <div>
      This is ExpensifyDashboardComponent
      <ExpenseListFilter/>
      <ExpensesList/>
    </div>
)

export default ExpensifyDashboard;