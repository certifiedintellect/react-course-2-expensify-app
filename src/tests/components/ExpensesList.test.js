import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesList} from '../../components/ExpensesList';
import moment from 'moment';

const expenses = [
    {
    id: '1',
    description: 'Gum',
    note: '',
    amount: 195,
    createdAt: moment(0)
  },{
      id: '2',
      description: 'Rent',
      note: '',
      amount: 195000,
      createdAt: moment(0).subtract(4, 'days').valueOf()
  }, {
     id: '3',
    description: 'Credit Card',
    note: '',
    amount: 4500,
    createdAt: moment(0).add(4, 'days').valueOf()
  }]
  

test('testing rendering of expenses list component ', () => {
    const wrapper = shallow(<ExpensesList expenses = {expenses}/>)
    expect(wrapper).toMatchSnapshot()
})


test('testing rendering of expenses list component with no expenses ', () => {
    const wrapper = shallow(<ExpensesList expenses = {[]}/>)
    expect(wrapper).toMatchSnapshot()
})
