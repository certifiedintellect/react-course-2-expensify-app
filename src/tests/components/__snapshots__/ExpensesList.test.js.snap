// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing rendering of expenses list component  1`] = `
<div>
  <h3>
    Expenses List
  </h3>
  <ExpenseItem
    amount={195}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseItem
    amount={195000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`testing rendering of expenses list component with no expenses  1`] = `
<div>
  <h3>
    Expenses List
  </h3>
</div>
`;
